File: scripts/bluesky_devto_notifier.js
// Bun has a native fetch API, no need for node-fetch
import { appendFile } from "node:fs/promises";

const DEVTO_API_URL = `https://dev.to/api/articles?per_page=1&username=${process.env.DEVTO_USER}`;
const LAST_ARTICLE_ID = process.env.LAST_ARTICLE_ID;

async function getLatestDevtoArticle() {
  try {
    const response = await fetch(DEVTO_API_URL);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    const articles = await response.json();
    if (articles?.length > 0) {
      return articles[0];
    }
  } catch (error) {
    console.error(`Error fetching articles from Dev.to API: ${error}`);
    process.end(1);
  }
}

const constructMessage = ({ title, description, url }) =>
  `${title}

${description}

${url}`;

export async function main() {
  try {
    const latestArticle = await getLatestDevtoArticle();

    if (!latestArticle) {
      throw new Error("No latest article found from Dev.to API.");
    }

    const {
      id,
      title,
      url,
      description = "No description available.",
    } = latestArticle;

    if (id.toString() === LAST_ARTICLE_ID) {
      console.log("No new articles to post to Bluesky.");
      return;
    }

    // Construct the message for Bluesky
    const message = constructMessage({ title, description, url });
    console.log(`Bluesky Message: ${message}`);

    // The actual posting to Bluesky will be handled by the GitHub Action
    // This script just prepares the message
    if (process.env.GITHUB_OUTPUT) {
      await appendFile(
        process.env.GITHUB_OUTPUT,
        `bluesky_message=${message}
`,
      );
      await appendFile(
        process.env.GITHUB_OUTPUT,
        `new_article_id=${id}
`,
      );
    }
  } catch (error) {
    console.error("An error occurred:", error);
    process.exit(1);
  }
}

File: scripts/bluesky_devto_notifier.py
import requests
import json
import os

DEVTO_API_URL = "https://dev.to/api/articles?per_page=1&username=" + os.getenv('DEVTO_USER')

def get_latest_devto_article():
    try:
        response = requests.get(DEVTO_API_URL)
        response.raise_for_status()
        articles = response.json()
        if articles:
            return articles[0]
    except requests.exceptions.RequestException as e:
        print(f"Error fetching articles from Dev.to API: {e}")
    return None

def main():
    latest_article = get_latest_devto_article()

    if latest_article:
        title = latest_article.get('title', 'Untitled Article')
        url = latest_article.get('url')
        description = latest_article.get('description', 'No description available.')

        # Construct the message for Bluesky
        # Bluesky posts are typically short, so combine title, description, and URL
        message = f"{title}\n\n{description}\n\n{url}"
        # Replace newlines with spaces for GITHUB_OUTPUT compatibility
        single_line_message = message.replace('\n', ' ').replace('\r', ' ')
        print(f"Bluesky Message: {message}")
        # The actual posting to Bluesky will be handled by the GitHub Action
        # This script just prepares the message
        with open(os.environ['GITHUB_OUTPUT'], 'a') as fh:
            print(f"bluesky_message={message}", file=fh)
    else:
        print("No latest article found from Dev.to API.")
        with open(os.environ['GITHUB_OUTPUT'], 'a') as fh:
            print("bluesky_message=", file=fh)

if __name__ == "__main__":
    main()
